lib/AST/Decl.cpp

```diff
@@ -856,18 +856,21 @@ GenericParamList *GenericContext::getGenericParams() const {
 }
 
 GenericSignature *GenericContext::getGenericSignature() const {
-  if (GenericSig)
-    return GenericSig;
+  return evaluateOrDefault(
+      getASTContext().evaluator,
+      GenericSignatureRequest{const_cast<GenericContext *>(this)}, nullptr);
+}
 
+llvm::Expected<GenericSignature *>
+GenericSignatureRequest::evaluate(Evaluator &evaluator,
+                                  GenericContext *GC) const {
   // The signature of a Protocol is trivial (Self: TheProtocol) so let's compute
   // it.
-  if (auto PD = dyn_cast<ProtocolDecl>(this)) {
+  if (auto PD = dyn_cast<ProtocolDecl>(GC)) {
     auto self = PD->getSelfInterfaceType()->castTo<GenericTypeParamType>();
     auto req =
         Requirement(RequirementKind::Conformance, self, PD->getDeclaredType());
-    const_cast<GenericContext *>(this)->GenericSig
-      = GenericSignature::get({self}, {req});
-    return GenericSig;
+    return GenericSignature::get({self}, {req});
   }
 
```

Comment:
This should be moved to TypeCheckGeneric.cpp, and you can handle NominalTypeDecl, TypeAliasDecl, ExtensionDecl, AbstractFunctionDecl and SubscriptDecl there too by calling into the existing code.